module/os/linux/spl/spl-kmem-cache.c

```diff
@@ -160,16 +160,34 @@ taskq_t *spl_kmem_cache_taskq;		/* Task queue for aging / reclaim */
 
 static void spl_cache_shrink(spl_kmem_cache_t *skc, void *obj);
 
+typedef struct {
+	uint32_t magic;
+	int real_allocation_size;
+} kvalloc_endmarker_t;
+const uint32_t kvmagic_alive = 0x11111111U;
+const uint32_t kvmagic_poison = 0xDEADDEADU;
+
 static void *
 kv_alloc(spl_kmem_cache_t *skc, int size, int flags)
 {
-	gfp_t lflags = kmem_flags_convert(flags);
-	void *ptr;
-
-	ptr = spl_vmalloc(size, lflags | __GFP_HIGHMEM);
+	BUILD_BUG_ON(!ISP2(PAGE_SIZE));
 
-	/* Resulting allocated memory will be page aligned */
-	ASSERT(IS_P2ALIGNED(ptr, PAGE_SIZE));
+	gfp_t lflags = kmem_flags_convert(flags);
+	void *ptr = NULL;
+
+	const int pad_size = sizeof (kvalloc_endmarker_t) + (PAGE_SIZE - 1);
+	char *unaligned_ptr = spl_kvmalloc(size + pad_size, lflags);
+	if (likely(unaligned_ptr != NULL)) {
+		ptr = (void*)P2ROUNDUP_TYPED(unaligned_ptr,
+		    PAGE_SIZE, uintptr_t);
+		kvalloc_endmarker_t *mkr = (kvalloc_endmarker_t *)
+		    ((char *)ptr + size);
+		mkr->magic = kvmagic_alive;
+		mkr->real_allocation_size = size;
```

Comment:
Maybe I'm misreading this, but I don't follow what's meant by "real" here.  Is this supposed to be the size that was requested (which it seems to be), or the size that we allocated from spl_kvmalloc() (which I would think is `size + pad_size`)?

Reply:
It's the size allocated from spl_kvmalloc() - we track it because we need to feed it back into spl_kmem_free_impl() (even though Linux spl_kmem_free_impl() doesn't use it!) and more importantly it lets us easily derive the original pointer we got from spl_kvmalloc() which we need to use for spl_kmem_free_impl().

Since that's unclear I welcome suggestions for better naming!  'underlying_allocation_size'?

Reply:
Nope, I managed to confuse myself too, so I think there's a real problem.  I'll take a look.

Reply:
> It's the size allocated from spl_kvmalloc()

Maybe I'm misreading the code, but that isn't what I see:

```
	char *unaligned_ptr = spl_kvmalloc(size + pad_size, lflags);
```
So at this point the size allocated from spl_kvmalloc is `size + pad_size`, right?  And then later we do `mkr->real_allocation_size = size`.  I don't see `size` being changed in between those, so shouldn't it be `real_allocation_size = size + pad_size`?

Reply:
Yeah this really doesn't look right, well spotted - it's not really taking into account the alignment round-up and it's not the 'real allocation size' even if it did.

Reply:
Pushed a change - real_allocation_size is simply replaced with the round-up offset, added more comments and hopefully things are easier to understand.  Not to mention, now hopefully correct.